File: sdks/java/client-java-core/src/main/java/io/featurehub/client/FeatureHubConfig.java
package io.featurehub.client;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.function.Supplier;

public interface FeatureHubConfig {
  /**
   * What is the fully deconstructed URL for the server?
   */
  String getRealtimeUrl();

  String apiKey();

  String baseUrl();

  // start the service
  void init();

  /**
   * The API Key indicates this is going to be server based evaluation
   */
  boolean isServerEvaluation();

  /**
   * returns a new context using the default edge provider and repository
   *
   * @return a new context
   */
  ClientContext newContext();

  /**
   * Allows you to create a new context for the user.
   *
   * @param repository - this repository is for this call only, it is not remembered, you should set the repository
   *                   on repository() to make it the default.
   *
   * @param edgeService - this edgeService is for this call only, it is not remembered, you should set it on
   *                    edgeService() to make it the default
   * @return a new context
   */
  ClientContext newContext(FeatureRepositoryContext repository, Supplier<EdgeService> edgeService);

  static boolean sdkKeyIsClientSideEvaluated(String sdkKey) {
    return sdkKey.contains("*");
  }

  void setRepository(FeatureRepositoryContext repository);
  FeatureRepositoryContext getRepository();

  void setEdgeService(Supplier<EdgeService> edgeService);
  Supplier<EdgeService> getEdgeService();

  /**
   * Allows you to specify a readyness listener to trigger every time the repository goes from
   * being in any way not reaay, to ready.
   * @param readynessListener
   */
  void addReadynessListener(ReadynessListener readynessListener);

  /**
   * Allows you to specify an analytics collector
   *
   * @param collector
   */
  void addAnalyticCollector(AnalyticsCollector collector);

  /**
   * Allows you to register a value interceptor
   * @param allowLockOverride
   * @param interceptor
   */
  void registerValueInterceptor(boolean allowLockOverride, FeatureValueInterceptor interceptor);

  /**
   * Allows you to query the state of the repository's readyness - such as in a heartbeat API
   * @return
   */
  Readyness getReadyness();

  /**
   * Allows you to override how your config will be deserialized when "getJson" is called.
   *
   * @param jsonConfigObjectMapper - a Jackson ObjectMapper
   */
  void setJsonConfigObjectMapper(ObjectMapper jsonConfigObjectMapper);
}

File: sdks/typescript/featurehub-javascript-client-sdk/app/feature_hub_config.ts
import { EdgeService } from './edge_service';
import { ClientContext } from './client_context';
import { InternalFeatureRepository } from './internal_feature_repository';
import { AnalyticsCollector } from './analytics';
import { FeatureStateValueInterceptor } from './interceptors';
import { FeatureHubRepository, Readyness, ReadynessListener } from './featurehub_repository';

// eslint-disable-next-line no-use-before-define
export type EdgeServiceProvider = (repository: InternalFeatureRepository, config: FeatureHubConfig) => EdgeService;
export type EdgeServiceSupplier = () => EdgeService;

export type FHLogMethod = (...args: any[]) => void;

export class FHLog {
  public static fhLog = new FHLog();

  public log: FHLogMethod = (...args: any[]) => {
    console.log(args);
  };

  public error: FHLogMethod = (...args: any[]) => {
    console.error(args);
  };

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public trace: FHLogMethod = (...args: any[]) => {};

  public quiet(): void {
    FHLog.fhLog.log = () => {
    };
    FHLog.fhLog.error = () => {
    };

    FHLog.fhLog.trace = () => {
    };
  }

  /**
   * @deprecated The method is deprecated. Use quiet() instead.
   */
  public Замолчи(): void {
    this.quiet();
  }
}

export const fhLog = FHLog.fhLog;

export interface FeatureHubConfig {
  readyness: Readyness;

  url(): string;

  // enable you to override the repository
  repository(repository?: FeatureHubRepository): FeatureHubRepository;

  // allow you to override the edge service provider
  edgeServiceProvider(edgeService?: EdgeServiceProvider): EdgeServiceProvider;

  // create a new context and allow you to pass in a repository and edge service
  newContext(repository?: FeatureHubRepository, edgeService?: EdgeServiceProvider): ClientContext;

  // is the repository client-side evaluated?
  clientEvaluated(): boolean;

  // add another API key
  apiKey(apiKey: string): FeatureHubConfig;

  // what are the API keys?
  getApiKeys(): Array<string>;

  // what is the host?
  getHost(): string;

  // initialize the connection outside of the creation of a context
  init(): FeatureHubConfig;

  // close any server connections
  close(): void;

  // add a callback for when the system is ready
  addReadynessListener(listener: ReadynessListener): void;

  // add an analytics collector
  addAnalyticCollector(collector: AnalyticsCollector): void;

  // add a value interceptor (e.g. baggage handler)
  addValueInterceptor(interceptor: FeatureStateValueInterceptor): void;
}
